### ISSUE TITLE
How to add custom fields

### ISSUE THREAD

**Issue by mobby38** [rfoltyns/log4j2-elasticsearch#26]
Hi, 

Would it be possible to add custom field value only in the Json message to elastic?
For example, it would be good to add the name or ip of the host, so that you can recognize where comes the log.

Thanks.

**Comment by rfoltyns** [comment]
@mobby38 You can use `ThreadContext` as decribed [here](https://github.com/rfoltyns/log4j2-elasticsearch/issues/18#issuecomment-488107973)

**Comment by mobby38** [comment]
Yes, but how you can do that without modifying the existing code.

The proof of concept I'm trying to achieve is to replace filebeat with just a log4j2 appender, but it misses those data, and I can't modify the code of the application.

What I could do is to provide Environment properties and give it to Log4j2 via the configuration.
Is there a way to do so?

**Comment by rfoltyns** [comment]
You can do that using configuration. Just switch to Log4j2 `JsonLayout` and do that with `KeyValuePair`s and env variables. You'll sacrifice the ability to use object pooling and customisable output though. Plus, you'll need to specify a new date column in your mappings as default set of mixins in `JsonLayout` serialize LogEvent a bit differently.
I'll add support for `KeyValuePair`s in `JacksonJsonLayout` in 1.4.

**Comment by mobby38** [comment]
Thanks :)

**Comment by mobby38** [comment]
In my opinion, you should also remove jsonIgnore in LogEventJacksonJsonMixIn and avoid renaming threadName by thread.
Just let the trick on timeMillis.

It tooks me some times to understand why not all attribute (including context) was hidden in the Json message.

**Comment by rfoltyns** [comment]
Good point on the `contextData`. There's a lot of confusion over this Log4j2 feature and I think I've gone too far while making the LogEvent 'thinner'.

Thankfully, you can augment any part of the output quite easily with `JacksonMixIn`.

**Comment by ricsil** [comment]
Is 1.4 coming?

**Comment by rfoltyns** [comment]
Yes. 1.4 should be out soon. It got heavily delayed due to `FileBackedRetryFailoverPolicy` work - sorry for that..

**Comment by ricsil** [comment]
Could you add an example of JsonLayout to docs? Simply replacing JacksonJsonLayout with JsonLayout doesn't error out but doesn't push anything to Elastic either, was wondering whether I missed something.

**Comment by rfoltyns** [comment]
@ricsil Sure, will do. Could you post your updated configuration in the meantime?

**Comment by ricsil** [comment]
So, this works fine (mix-in is there for timestamp override)
````
    <Elasticsearch name="elasticsearchAsyncBatch">
      <RollingIndexName indexName="test-log-app" pattern="yyyy-MM-dd" />
      <JacksonJsonLayout>
        <JacksonMixIn mixInClass="path.to.utils.CustomLogEventMixIn"
          targetClass="org.apache.logging.log4j.core.LogEvent"/>
      </JacksonJsonLayout>
      <AsyncBatchDelivery>
        <JestHttp serverUris="http://elastic-server:9200" />
      </AsyncBatchDelivery>
    </Elasticsearch>
````

Replacing JacksonJson with Json doesn't produce anything
````
    <Elasticsearch name="elasticsearchAsyncBatch">
      <RollingIndexName indexName="test-log-app" pattern="yyyy-MM-dd" />
      <JsonLayout />
      <AsyncBatchDelivery>
        <JestHttp serverUris="http://elastic-server:9200" />
      </AsyncBatchDelivery>
    </Elasticsearch>
````

On the other hand, replacing JacksonJsonLayout with EcsLayout (not forgetting to add dependency) works fine:
````
    <Elasticsearch name="elasticsearchAsyncBatch">
      <RollingIndexName indexName="test-log-app" pattern="yyyy-MM-dd" />
      <EcsLayout />
      <AsyncBatchDelivery>
        <JestHttp serverUris="http://elastic-server:9200" />
      </AsyncBatchDelivery>
    </Elasticsearch>
````

https://github.com/elastic/ecs-logging-java/tree/master/log4j2-ecs-layout

**Comment by rfoltyns** [comment]
@ricsil you need
`<JsonLayout compact="true"/>` cause default one prints pretty JSON. Have a look and [updated docs](https://github.com/rfoltyns/log4j2-elasticsearch/tree/master/log4j2-elasticsearch-core#log4j2-jsonlayout)

**Comment by rfoltyns** [comment]
`VirtualProperty` support released in 1.4.
